### ISSUE TITLE
Proposed Dockerfile

### ISSUE THREAD

**Issue by AndrewCSQ** [pyro-ppl/numpyro#992]
Hi numpyro team,

I'm building a package on top of numpyro, and hoping to be able to standardize the delivery GPU-acceleration on jax using docker images. jax itself does not provide docker images currently. I've observed that the tensorflow-gpu images work, and opened an issue on jax separately for a jaxlib-based docker image.

The dockerfile I'm currently using is roughly so

```
FROM nvidia/cuda:11.2.2-cudnn8-devel-ubuntu20.04

# declare the image name
ENV IMG_NAME=11.2.2-cudnn8-devel-ubuntu20.04 \
    # declare what jaxlib and jax versions to use
    # if a CI/CD system is expected to pass in these arguments
    # the dockerfile should be modified accordingly
    JAXLIB_CUDA=112 \
    JAXLIB_VERSION=0.1.62 \
    JAX_VERSION=0.2.10
    NUMPYRO_VERSION=0.6.0

RUN apt update && \
    apt install python3-dev python3-pip python3-setuptools -y

# install python packages via pip
RUN pip3 install --user \
    numpyro==${NUMPYRO_VERSION} \
    jax==${JAX_VERSION} \
    jaxlib==${JAXLIB_VERSION}+cuda${JAXLIB_CUDA} -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
I'm not sure what sort of CI/CD system numpyro has (I've looked briefly into version numbering and pinning for release) - so the image will definitely need to be modified accordingly (for example, upon release when the image is built jax, jaxlib, numpyro versions will need to be passed in as environment variables or the dockerfile dynamically generated).

I'm prepared to step up and maintain the dockerfile for a period of time if the team is amenable to it.

**Comment by fehiepsi** [comment]
Hi @AndrewCSQ, I just have little experience with Docker so I might not provide some useful thoughts. Please feel free to adopt the convention given by the JAX team here. (I thought Dockerfile is used to simplify the process to build some package from its source - when it is tricky to use `pip install` - but it seems not to be the case here.) Do you intend to make it work across platforms/devices? FYI, Pyro also has an old [docker folder](https://github.com/pyro-ppl/pyro/tree/dev/docker) that you can adopt here. Regarding versioning, we are going to "not ping" jax/jaxlib versions for the next release so it would be nice to have an automatic solution here (probably a solution is to build jaxlib from source?).

**Comment by AndrewCSQ** [comment]
Hi @fehiepsi, thanks for responding to this! I think I should clarify what the use case is, and what I need to make this "automatic". I'll then talk about why I largely disregarded the Pyro docker folder, and some thoughts about cross-platform (short version: Linux and Windows, limited by nvidia-docker requirements).

### Pre-Requisites to Use the Docker Image (resulting from the Dockerfile)

Users would need a Linux (or WSL, see note below on platform compatibility) distribution with the Nvidia driver (CUDA is not required), and Docker of course (Moby may be tested against, podman and kubernetes will definitely _not_ be tested against by me). They would then need to install [nvidia-docker](https://github.com/NVIDIA/nvidia-docker), and they're off to the races.

### Docker Use Case

One of the use cases of Docker is indeed to provide a consistent and "clean" build environment for building packages from source. But the same need for a consistent and "clean" environment applies to _running_ code also. In particular, to use GPU-accelerated JAX, one needs to be on Ubuntu (16.04 or later), and run particular combinations of CUDA+CUDNN (according to the wheel chosen). Think of it as an "OS-wide venv". Given the integration that Docker has with CI/CD platforms, I think it will also make it easier for testing (especially in GPU situations like multi-GPU).

Getting Ubuntu 16.04 or later with the right CUDA+CUDNN is not always possible (what if I have to run pytorch on the same computer, and that wants a different CUDA+CUDNN version that's incompatible with the driver needed for CUDA accelerated JAX? What if my organization says we're using RHEL instead? Then all the `*.so` files will be in unexpected places). Docker allows the user to have exactly the environment they need to use numpyro. This is particularly useful for numpyro because numpyro practices version pinning upon release for JAX/JAXLIB (understandably, until the API stabilizes - I've read the file).

In practice, the Dockerfile/image is meant for the developer/user to "layer" their needs on top. If I write a package on top of numpyro, what I would do is "pull" the Docker image here, then add `pip install <my_cool_package>` on top of it, and re-package it for my needs. As a user, I would write a project-specific Dockerfile (so that my colleagues can replicate my work cleanly, even years down the line), for example adding Jupyter if I use notebooks, building on top of the numpyro distributed image.

### What I Need to Automate

When you say that you are going to "not ping" jax/jaxlib version do you mean that you will not require a specific version of jax/jaxlib version for the next release (unlike 0.6.0 which requires jax 0.2.10 and jaxlib 0.1.62?) If that is the case, then automation should be the "latest tag of jax/jaxlib". In an ideal world where jax starts providing images of their own, then we can just pull the "latest stable" version of the jax/jaxlib image, layer numpyro on top of it, and we are done. (In this case, we are the "developer" on top of the image in the use-case above). If jax/jaxlib does not provide their own images, then there will have to be a script every numpyro release to use the latest jax/jaxlib tags, and test that nothing breaks.

In the case where numpyro will continue to specify particular jax/jaxlib versions, then I will need to hook into the existing version pinning scripts (that I will need the team's help with). This will probably be in the form of a "Dockerfile generating script" that outputs basically the file above, but dynamically sets the JAXLIB_VERSION etc. environment variables.

### Existing Pyro Docker Folder, Cross-Platform?

The existing Pyro Docker folder is too bloated. It installs miniconda (which is nice, but not necessary), and does everything through anaconda repositories. Since numpyro is not in the base anaconda repository (one needs to use conda-forge instead), this may not be desirable. A direct `pip install` is preferable for the Docker use-case (in my opinion, of course). It also depends on pytorch bundling the whole CUDA+CUDNN stack inside the package itself (which JAX has stated they will not do). It also creates a user (pyromancer) to install all python packages. This is useful as an end-user image (because users should not be root inside containers that they are running) but less useful for developers (my package on top of numpyro should not have to have the "pyromancer" user just to run numpyro).

My approach here is to take a clean nvidia-cudnn-devel image on Ubuntu (20.04), and add python, then pip install only what I need. Keeps the dependencies low, and increases maintainability.

Platform support is [dictated by what nvidia-docker gives](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#platform-support). Right now, that's Linux only, with the possibility of [experimental WSL2 support](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#installing-wip). It should be noted that users that do manage to get WSL2 working can expect ~60% performance. That's not a part of the stack I'm interested in maintaining, so my compatibility target is "whatever nvidia-docker says will work".

**Comment by fehiepsi** [comment]
Thanks for explaining! The usage case makes sense to me. As long as it is not straightforward to set up an environment, an automatic solution is very handy. (I don't have enough experience with Dockerfile but I have seen many people mentioned it.)

> If jax/jaxlib does not provide their own images, then there will have to be a script every numpyro release to use the latest jax/jaxlib tags, and test that nothing breaks.

Very nice idea! Let's see if we'll have a docker file in JAX. If not, we can always get the latest jaxlib version by grepping PyPI releases. Regarding testing, I'm not sure what's the best practice to reduce maintaining overhead. Please feel free to stick with your solution.

> Preferably No Build From Source

Sure, this is already nice for the first version. We can iteratively improve the Docker file, e.g. to include build instructions for Windows if this is possible in the future.

Regarding Pyro docker, sorry for making confusion. I just wanted to let you know a structure that you can mimic here (e.g. putting Dockerfile, README.md, and probably Makefile to customize the installation,... in the `docker` folder). Please ignore the implementation details.

**Comment by AndrewCSQ** [comment]
Thank you for the feedback! Here's my proposal:

1. I will open a PR later in the week (swamped with other work right now) to create a docker folder in the repository for the Dockerfile. 
2. I can also draft documentation for how a user can make the docker image (I'm a bit unfamiliar with sphinx, so you will need to guide me on how to edit what `.rst` files)

In terms of implementation, right now I will target numpyro 0.6.0, which has manually specified jax and jaxlib versions. No customization will be supported.

What I'm hoping for is:

1. Maybe the numpyro team can consider setting up a [Dockerhub Account](https://docs.docker.com/docker-hub/) to distribute the image? That way users can just use `docker pull numpyro/numpyro-gpu:latest` instead of needing to build the docker image from the repository.
    - I don't know what the relationship is between numpyro, the original pyro team, and pytorch, so maybe if there's already an existing account we can leverage that
 2. Some guidance on how to run the tests. For now, I can run the tests in the repo every release using the Dockerfile to make sure that the tests run (on my consumer hardware single GPU system). Ideally, if there is support from the pyro team (or whatever company is backing them), they will automate running the tests inside the docker image prior to release.

**Comment by fehiepsi** [comment]
> unfamiliar with sphinx

You'll just need to use markdown in a `README.md` file. Writing a markdown is very similar to how to write comments on github. But I can help to review.

> some guidance on how to run the tests

Let me bring this to our next meeting (in a few weeks) but I don't think we can get support for a GPU testing system. Probably it is best to take further action based on the support from the JAX team (i.e. if they test their docker image, then it is unlikely that numpyro will break on GPU while working well on CPU). Personally, I prefer to thinking this as a feature maintained by the community and reducing the maintaining overhead for the current Pyro team.

**Comment by AndrewCSQ** [comment]
Testing the Docker image will not require the use of a GPU system (since CUDA compiled JAX works using just a CPU after all). If right now the testing procedure is:

- Clone `numpyro` repo on a test machine, run pytest, check that everything passes.

What I'm proposing is appending the step:

- Build (or pull) docker image, clone `numpyro` repo inside that image, run pytest, check that everything passes.

So basically adding the docker image as a "supported OS / environment" that gets tested against. It will require whatever existing testing mechanism to support Docker (but that shouldn't be too hard). If that's too onerous still, I will for the foreseeable future be running my own tests (but my commitment to maintenance may not extend to extensive, comprehensive testing especially if the tests are compute intensive).

**Comment by fehiepsi** [comment]
> my commitment to maintenance may not extend to extensive

No worries, your help on some first iterations is already very valuable! I can add some "help wanted" label if someone raises an issue with Dockerfile. IIUC, with `pip` installation, the main issues will belong to numpyro or jax incompatibility, rather than the building instructions in the Dockerfile.

> will future numpyro versions require specific versions of jax and jaxlib also?

Yup, that's right.

**Comment by AndrewCSQ** [comment]
Closed with merging of #996
